From ab08942da5e2e67165976199e48b7e2eeb100e45 Mon Sep 17 00:00:00 2001
From: Shahed Shaikh <shahedx.shaikh@intel.com>
Date: Fri, 28 Aug 2020 10:55:16 +0800
Subject: [PATCH] DRV_DPM_SW-370: DPM enhance dp_queue_conf

- enhance if-else condition for  DP_QUEUE_DROP_WRED and
  DP_QUEUE_DROP_TAIL in dp_queue_conf_get() and dp_queue_conf_set()
- In dp_queue_conf_set() remove checks for flags field
- In dp qos debufs node, add option "set_q_conf" to change queue
  conf settings

Signed-off-by: Shahed Shaikh <shahedx.shaikh@intel.com>
---
 drivers/net/datapath/dpm/datapath_proc_qos.c       | 140 +++++++++++++++------
 .../net/datapath/dpm/gswip31/datapath_ppv4_api.c   |  29 +++--
 2 files changed, 116 insertions(+), 53 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath_proc_qos.c b/drivers/net/datapath/dpm/datapath_proc_qos.c
index a2884379d362..0741b1bb7e7d 100644
--- a/drivers/net/datapath/dpm/datapath_proc_qos.c
+++ b/drivers/net/datapath/dpm/datapath_proc_qos.c
@@ -944,6 +944,15 @@ void qos_create_qos_help(void)
 	pr_info("\n");
 	pr_info("   CLEAR STATS: echo clear > %s\n", DEBUGFS_QOS);
 	pr_info("\n");
+	pr_info("   PRINT QUEUE CFG: echo q_conf <id> > %s\n", DEBUGFS_QOS);
+	pr_info("   SET QUEUE CFG: echo set_q_conf <id> [[options] [value]] > %s\n",
+		DEBUGFS_QOS);
+	pr_info("       options : action [0x1 : disable, 0x2 : enable, 0x4 suspend]\n");
+	pr_info("                        [0x8 : resume]\n");
+	pr_info("               : drop [0 : tail drop mode, 1 : wred mode]\n");
+	pr_info("               : green_min_size/green_max_size/\n");
+	pr_info("               : green_slope/yellow_min_size/yellow_max_size/\n");
+	pr_info("               : yellow_slope/min_guaranteed/max_allowed\n");
 }
 
 static int get_arbi(struct dp_node_child *c, enum dp_arbitate *p)
@@ -1047,11 +1056,45 @@ static int add_qos_link(struct dp_qos_link *cfg, int flag)
 	return dp_link_add(cfg, flag);
 }
 
+static void print_queue_conf(struct dp_queue_conf *q_conf)
+{
+
+	if (q_conf->act & DP_NODE_EN)
+		pr_info("Q(%d) action is ENABLED:[%d]\n",
+			q_conf->q_id, q_conf->act);
+	else
+		pr_info("Q(%d) action is BLOCKED:[%d]\n",
+			q_conf->q_id, q_conf->act);
+	if (q_conf->drop == DP_QUEUE_DROP_WRED) {
+		pr_info("Q(%d) is in WRED MODE:[%d]\n",
+			q_conf->q_id, q_conf->drop);
+	} else {
+		pr_info("Q(%d) is in DROP TAIL MODE:[%d]\n",
+			q_conf->q_id, q_conf->drop);
+	}
+	pr_info("Q(%d) is green_min_size:[%d]\n",
+		q_conf->q_id, q_conf->min_size[0]);
+	pr_info("Q(%d) is green_max_size:[%d]\n",
+		q_conf->q_id, q_conf->max_size[0]);
+	pr_info("Q(%d) is yellow_min_size:[%d]\n",
+		q_conf->q_id, q_conf->min_size[1]);
+	pr_info("Q(%d) is yellow_max_size:[%d]\n",
+		q_conf->q_id, q_conf->max_size[1]);
+	pr_info("Q(%d) is max_allowed:[%d]\n",
+		q_conf->q_id, q_conf->wred_max_allowed);
+	pr_info("Q(%d) is min_gauranteed:[%d]\n",
+		q_conf->q_id, q_conf->wred_min_guaranteed);
+	pr_info("Q(%d) is green_wred_slope:[%d]\n",
+		q_conf->q_id, q_conf->wred_slope[0]);
+	pr_info("Q(%d) is yellow_wred_slope:[%d]\n",
+		q_conf->q_id, q_conf->wred_slope[1]);
+}
+
 ssize_t proc_qos_write(struct file *file, const char __user *buf,
 		       size_t count, loff_t *ppos)
 {
-	char str[100];
-	char *param_list[16];
+	char str[200];
+	char *param_list[25];
 	int num;
 	size_t len;
 
@@ -1455,42 +1498,63 @@ ssize_t proc_qos_write(struct file *file, const char __user *buf,
 			pr_err("dp_queue_conf_get failed\n");
 			return count;
 		}
-		if (q_conf.act & DP_NODE_EN)
-			pr_info("Q(%d) action is ENABLED:[%d]\n",
-				q_conf.q_id, q_conf.act);
-		else
-			pr_info("Q(%d) action is BLOCKED:[%d]\n",
-				q_conf.q_id, q_conf.act);
-		if (q_conf.drop == DP_QUEUE_DROP_WRED) {
-			pr_info("Q(%d) is in WRED MODE:[%d]\n",
-				q_conf.q_id, q_conf.drop);
-			pr_info("Q(%d) is min_size_1:[%d]\n",
-				q_conf.q_id, q_conf.min_size[0]);
-			pr_info("Q(%d) is max_size_1:[%d]\n",
-				q_conf.q_id, q_conf.max_size[0]);
-			pr_info("Q(%d) is min_size_2:[%d]\n",
-				q_conf.q_id, q_conf.min_size[1]);
-			pr_info("Q(%d) is max_size_2:[%d]\n",
-				q_conf.q_id, q_conf.max_size[1]);
-			pr_info("Q(%d) is max_allowed:[%d]\n",
-				q_conf.q_id, q_conf.wred_max_allowed);
-			pr_info("Q(%d) is min_gauranteed:[%d]\n",
-				q_conf.q_id, q_conf.wred_min_guaranteed);
-			pr_info("Q(%d) is wred_slope_1:[%d]\n",
-				q_conf.q_id, q_conf.wred_slope[0]);
-			pr_info("Q(%d) is wred_slope_2:[%d]\n",
-				q_conf.q_id, q_conf.wred_slope[1]);
-		} else {
-			pr_info("Q(%d) is in DROP TAIL MODE:[%d]\n",
-				q_conf.q_id, q_conf.drop);
-			pr_info("Q(%d) is min_size_1:[%d]\n",
-				q_conf.q_id, q_conf.min_size[0]);
-			pr_info("Q(%d) is max_size_1:[%d]\n",
-				q_conf.q_id, q_conf.max_size[0]);
-			pr_info("Q(%d) is min_size_2:[%d]\n",
-				q_conf.q_id, q_conf.min_size[1]);
-			pr_info("Q(%d) is max_size_2:[%d]\n",
-				q_conf.q_id, q_conf.max_size[1]);
+
+		print_queue_conf(&q_conf);
+	} else if (dp_strncmpi(param_list[0], "set_q_conf",
+			       strlen("q_conf")) == 0) {
+		struct dp_queue_conf cfg = {0};
+		int i;
+
+		cfg.inst = 0;
+		cfg.q_id = dp_atoi(param_list[1]);
+		if (dp_queue_conf_get(&cfg, 0)) {
+			pr_err("dp_queue_conf_get failed\n");
+			return count;
+		}
+
+		i = 2; /* skip param 0 and param 1*/
+		while (i < num) {
+			if (dp_strncmpi(param_list[i], "action",
+					strlen("action")) == 0) {
+				cfg.act = dp_atoi(param_list[++i]);
+			} else if (dp_strncmpi(param_list[i], "drop",
+					strlen("drop")) == 0) {
+				cfg.drop = dp_atoi(param_list[++i]);
+			} else if (dp_strncmpi(param_list[i], "green_min_size",
+					strlen("green_min_size")) == 0) {
+				cfg.min_size[DP_COLOR_GREEN] = dp_atoi(param_list[++i]);
+			} else if (dp_strncmpi(param_list[i], "green_max_size",
+					strlen("green_max_size")) == 0) {
+				cfg.max_size[DP_COLOR_GREEN] = dp_atoi(param_list[++i]);
+			} else if (dp_strncmpi(param_list[i], "green_slope",
+					strlen("green_slope")) == 0) {
+				cfg.wred_slope[DP_COLOR_GREEN] = dp_atoi(param_list[++i]);
+			} else if (dp_strncmpi(param_list[i], "yellow_min_size",
+					strlen("yellow_min_size")) == 0) {
+				cfg.min_size[DP_COLOR_YELLOW] = dp_atoi(param_list[++i]);
+			} else if (dp_strncmpi(param_list[i], "yellow_max_size",
+					strlen("yellow_max_size")) == 0) {
+				cfg.max_size[DP_COLOR_YELLOW] = dp_atoi(param_list[++i]);
+			} else if (dp_strncmpi(param_list[i], "yellow_slope",
+					strlen("yellow_slope")) == 0) {
+				cfg.wred_slope[DP_COLOR_YELLOW] = dp_atoi(param_list[++i]);
+			} else if (dp_strncmpi(param_list[i], "min_guaranteed",
+					strlen("min_guaranteed")) == 0) {
+				cfg.wred_min_guaranteed = dp_atoi(param_list[++i]);
+			} else if (dp_strncmpi(param_list[i], "max_allowed",
+					strlen("max_allowed")) == 0) {
+				cfg.wred_max_allowed = dp_atoi(param_list[++i]);
+			} else {
+				pr_info("Wrong Parameter [%s]:\n", param_list[i]);
+				qos_create_qos_help();
+				return count;
+			}
+			i++;
+		}
+
+		if (dp_queue_conf_set(&cfg, 0)) {
+			pr_err("dp_queue_conf_get failed\n");
+			return count;
 		}
 	} else {
 		pr_info("Wrong Parameter:\n");
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
index 227d29ec8d34..cbb26d07a0f1 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
@@ -3131,28 +3131,27 @@ int dp_queue_conf_set_31(struct dp_queue_conf *cfg, int flag)
 		pr_err("qos_queue_conf_get fail:%d\n", cfg->q_id);
 		goto EXIT;
 	}
-	if (flag & (cfg->act & DP_NODE_DIS))
+	if (cfg->act & DP_NODE_DIS)
 		conf->blocked = 1;
-	else if (flag & (cfg->act & DP_NODE_EN))
+	else if (cfg->act & DP_NODE_EN)
 		conf->blocked = 0;
 
 	priv->qos_queue_stat[cfg->q_id].blocked = conf->blocked;
 
-	if (flag & (cfg->drop == DP_QUEUE_DROP_WRED)) {
+	if (cfg->drop == DP_QUEUE_DROP_WRED)
 		conf->wred_enable = 1;
-		conf->wred_max_avg_green = cfg->max_size[0];
-		conf->wred_max_avg_yellow = cfg->max_size[1];
-		conf->wred_min_avg_green = cfg->min_size[0];
-		conf->wred_min_avg_yellow = cfg->min_size[1];
-		conf->wred_slope_green = cfg->wred_slope[0];
-		conf->wred_slope_yellow = cfg->wred_slope[1];
-		conf->wred_min_guaranteed = cfg->wred_min_guaranteed;
-		conf->wred_max_allowed = cfg->wred_max_allowed;
-	} else if (flag & (cfg->drop == DP_QUEUE_DROP_TAIL)) {
+	else if (cfg->drop == DP_QUEUE_DROP_TAIL)
 		conf->wred_enable = 0;
-		conf->wred_min_guaranteed = cfg->wred_min_guaranteed;
-		conf->wred_max_allowed = cfg->wred_max_allowed;
-	}
+
+	conf->wred_max_avg_green = cfg->max_size[0];
+	conf->wred_max_avg_yellow = cfg->max_size[1];
+	conf->wred_min_avg_green = cfg->min_size[0];
+	conf->wred_min_avg_yellow = cfg->min_size[1];
+	conf->wred_slope_green = cfg->wred_slope[0];
+	conf->wred_slope_yellow = cfg->wred_slope[1];
+	conf->wred_min_guaranteed = cfg->wred_min_guaranteed;
+	conf->wred_max_allowed = cfg->wred_max_allowed;
+
 	if (dp_qos_queue_set(priv->qdev, node_id, conf)) {
 		pr_err("failed to qos_queue_set:%d\n", cfg->q_id);
 		goto EXIT;
